keps/sig-cloud-provider/0013-build-deploy-ccm.md

```diff
@@ -0,0 +1,494 @@
+in K8s/K8s for each cloud-provider. 
+
+The majority of the controller-manager framework and its dependencies also needs to be shared. This is code that would 
+be shared even after cloud providers are removed from K8s/K8s. As such it probably makes sense to make the controller 
+manager framework its own K8s/K8s Staging repo. 
+
+It should be generally possible for cloud providers to determine where a controller runs and even over-ride specific 
+controller functionality. Please note that if a cloud provider exercises this possibility it is up to that cloud provider 
+to keep their custom controller conformant to the K8s/K8s standard. This means any controllers may be run in either KCM 
+or CCM. As an example the NodeIpamController, will be shared acrosss K8s/K8s and K8s/cloud-provider-gce, both in the 
+short and long term. Currently it needs to take a cloud provider to allow it to do GCE CIDR management. We could handle 
+this by leaving the cloud provider interface with the controller manager framework code. The GCE controller manager could 
+then inject the cloud provider for that controller. For everyone else (especially the KCM) NodeIpamController is 
+interesting because currently everyone needs its generic behavior for things like ranges. However Google then wires in 
+the cloud provider to provide custom functionality on things like CIDRs. The thought then is that in the short term we 
+allow it to be run in either KCM or CCM. Most cloud providers would run it in the KCM, Google would run it in the CCM. 
+When we are ready to move cloud provider code out of K8s/K8s, we remove the cloud provider code from the version which 
+is in K8s/K8s and continue to have flags to control if it runs. K8s/Cloud-Provider-Google could then have an enhanced 
+version which is run in the CCM. Other controllers such as Route and Service needs to run in either the KCM or CCM. For 
+things like K8s/K8s e2e tests we will always want these controllers in the K8s/K8s repo. Having it in the K8s/K8s repo 
+is also useful for keeping the behavior of these sort of core systems consistent.
+
+#### Staging
+
+There are several sections of code which need to be shared between the K8s/K8s repo and the K8s/Cloud-provider repos. 
+The plan for doing that sharing is to move the relevant code into the Staging directory as that is where we share code 
+today. The current Staging repo has the following packages in it.
+- Api
+- Apiextensions-apiserver
+- Apimachinery
+- Apiserver
+- Client-go
+- Code-generator
+- Kube-aggregator
+- Metrics
+- Sample-apiserver
+- Sample-Controller
+
+With the additions needed in the short term to make this work; the Staging area would now need to look as follows. 
+- Api
+- Apiextensions-apiserver
+- Apimachinery
+- Apiserver
+- Client-go
+- **Controller**
+  - **Cloud**
+  - **Service**
+  - **NodeIpam**
+  - **Route**
+  - **?Volume?**
+- **Controller-manager**
```

Comment:
Been thinking about this more lately. 

I think we should be consistent with what is happening with `kube-apiserver` & `kube-controller-manager`. From what I can tell it seems like it's unclear whether all the Kubernetes components will be broken up into its own repos (unless this was stated somewhere and I missed it). Until we know for sure, I think we should hold off on trying to move controller-manager and controllers code. 

I also think it'll be easier to get buy-in if we scope this KEP for only moving cloud provider code and worry about controllers when the story around that is clearer.

Reply:
I forgot about https://github.com/kubernetes/community/pull/2354 but it's unclear if `k8s.io/{component}` repos are for component types AND the actual code for the component. Will check and report back.

Reply:
nvm, I asked the sig cluster lifecycle folks, all the code will be moved to k8s.io/components

Reply:
Leaving controllers where they're at now means they can be used in-tree and still be vendored in K8s/CP. Not saying that I don't want to see controllers be pulled out of K8s/K8s but it may be easier to move provider code out first and move controllers when we have a better idea of what that would look like (also to reduce # of high impact changes in the k8s/k8s codebase). I'm a bit on the fence about this since I don't know what the implications of moving controllers (cloud node controller, service, routes, etc) into staging would be 🤔.

Reply:
So if leave them where they are they can definitely be used in in-tree. However we have been trying for awhile to use staging as a delimiter for what can be vendored in externally. If we move the relevant controllers to staging they can be used in-tree and they are then part of the code we have officially marked as being vendorable for K8s/CP. I would like to see this kept in the KEP as its the right thing to do. If we want to mark it as a longer term goal which is the right thing to do, I would be fine with that.

Reply:
Worth reading @smarterclayton thoughts on staging found at https://github.com/kubernetes/kubernetes/issues/28559. It references client-go but I think it generally applies to packages which need to be shared beyond K8s/K8s.